### ISSUE TITLE
LFN support in 32-Bit DOS / Protected mode / DOS4G environment broken

### ISSUE THREAD

**Issue by oerg866** [open-watcom/open-watcom-v2#1009]
Hello,

I'm trying to use long file names in a protected mode environment. 

Take this sample code:

```
#include <stdio.h>

int main() {
    int file = _open("C:\\dshfisdlfgjsldfjhsjldgfhsl", O_CREAT | O_TRUNC | O_BINARY);
    _close(file);
}
```

Compiled with `-D__WATCOM_LFN__ wcl386 -3 -bt=dos4g -l=dos4g main.c`
The directory listing of C:\ yields:

```
DSHFISDL                                        0  12-01-22  6:33p  DSHFISDL
```

Whereas the same program compiled in real mode (`wcl386 -D__WATCOM_LFN__ -3 -bt=dos main.c`) yields:

```
DSHFIS~1                                        0  12-01-22  6:33p  dshfisdlfgjsldfjhsjldgfhsl
```

Tested with clean DOS 7.10 / Win98 boot disk with HIMEM.SYS and DOSLFN 0.41f.

It would be nice to see this fixed.

**Comment by jmalak** [comment]
It is issue with incorrect option for compile and link utility.
if you use wcl386 utility it generate 32-bit protected mode code.
Take into account that if you compile and link for 32-bit DOS Extended code you need specify compiler target as DOS and select what DOS Extender you want to use. It is difference if you want compile for 16-bit real mode code then you compile and klink for DOS only.

Try following to test how it works.
**16-bit DOS executable**
`wcl -bcl=dos -D__WATCOM_LFN__ main.c`
or
`wcl -bt=dos -l=dos -D__WATCOM_LFN__ main.c`

**32-bit DOS executable with DOS4GW Extender**
`wcl386 -bt=dos -l=dos4g -D__WATCOM_LFN__ main.c`
**32-bit DOS executable with Causeway Extender**
`wcl386 -bt=dos -l=causeway -D__WATCOM_LFN__ main.c`

Anyway you must always specify -bt=dos for DOS.
Take into acount that -bcl=xx option is as if you specify -bt=xx -l=xx
This is invalid for 32-bit DOS with DOS extender because you need specify which one you want to use.

**Comment by oerg866** [comment]
Yes, this is how I compiled the protected mode example.

This is about difference in behavior from compiled programs (16-Bit Real Mode vs 32-Bit Protected Mode).

Observe the following two outputs:

![image](https://user-images.githubusercontent.com/1124157/205179092-abbee37e-2769-484c-9601-15870c305001.png)
![image](https://user-images.githubusercontent.com/1124157/205179065-2cc98b07-02a5-4c6f-bb40-6eb9fda0de0e.png)


Upper was compiled with `wcl -D__WATCOM_LFN__ -3 -bt=dos -fe=realmode.exe main.c`
Lower was compiled with `wcl386 -D__WATCOM_LFN__ -3 -bt=dos4g -l=dos4g -fe=protmode.exe main.c`

Best
Eric

**Comment by jmalak** [comment]
But both compile and build commands are wrong.
Correct should be
Upper was compiled with wcl386 -D__WATCOM_LFN__ -3 -bt=dos -l=dos4g -fe=protmode.exe main.c
Lower was compiled with wcl -D__WATCOM_LFN__ -3 -bt=dos -l=dos -fe=realmode.exe main.c

**Comment by oerg866** [comment]
Exactly the same result. 

Sidenote, according to the internet, `-bt=dos4g` is synonymous for `-bt=dos` when liking a protected mode application.

**Comment by oerg866** [comment]
By the way, the same faulty behavior happens with `fopen`, too

**Comment by jmalak** [comment]
I don't know where you get information that -bt=dos4g is correct.
No it is not true only -bt=dos is correct because it defines proper predefined macros.
I tested it on Windows XP with DOSX extender and there is file created with LFN for both 16 and 32-bit DOS code.
But it is irrelevant for your environment.

What about created file attributes on your system?

**Comment by jmalak** [comment]
anyway your combination surprise me
DOS 7.10 / Win98 + DOSLFN 0.41f
Up to now I was thinking that DOS for Win98 contains LFN support.

**Comment by oerg866** [comment]
This is a boot floppy, it's not running *on* windows. It's pure DOS 7.10.

Best
Eric

**Comment by jmalak** [comment]
Do you want to say that DOS 7.10 doesn't contains LFN functions and they are available only under Windows?

**Comment by Baron-von-Riedesel** [comment]
> Do you want to say that DOS 7.10 doesn't contains LFN functions and they are available only under Windows?

That's correct. If you want LFN in pure DOS 7.1, you'll have to install a driver, DOSLFN.COM, for example.

**Comment by jmalak** [comment]
OK.
I will review and check under this configuration.

**Comment by jmalak** [comment]
There is incorrect order of calling LFN and SFN functions.
It works on Windows XP, but not on pure DOS.
It is necessary reorder LFN and SFN functions to work with various environments.

**Comment by oerg866** [comment]
Thanks for checking, looking forward for fix! :)
